frontend/catalyst/device/op_support.py

```diff
@@ -40,10 +40,95 @@ def is_supported(op: Operator, capabilities: DeviceCapabilities) -> bool:
     return op.name in capabilities.operations
 
 
+def _is_grad_recipe_same_as_catalyst(op):
+    """Checks that the grad_recipe for the op matches the hard coded one in Catalyst."""
+
+    def _is_active(maybe_tracer):
+        return isinstance(maybe_tracer, jax.core.Tracer)
+
+    def _is_grad_recipe_active(grad_recipe):
+        active = False
+        for recipe in grad_recipe:
+            left, right = recipe
+            active_left = any(map(_is_active, left))
+            active_right = any(map(_is_active, right))
+            active |= active_left or active_right
+        return active
+
+    if _is_grad_recipe_active(op.grad_recipe):
+        # An active grad recipe is never the same as the one in catalyst
+        return False
+
+    if len(op.data) != len(op.grad_recipe):
+        return False
+
+    valid = True
+    for grad_recipe in op.grad_recipe:
+        left, right = grad_recipe
+        # exp_param_shift_rule_{left,right} are constants in Catalyst
+        # we must ensure that the rules seen in the op match Catalyst's implementation.
+        exp_param_shift_rule_left = np.array([0.5, 1.0, np.pi / 2])
+        exp_param_shift_rule_right = np.array([-0.5, 1.0, -np.pi / 2])
+        obs_param_shift_rule_left = np.array(left)
+        obs_param_shift_rule_right = np.array(right)
+        is_left_valid = np.allclose(obs_param_shift_rule_left, exp_param_shift_rule_left)
+        is_right_valid = np.allclose(obs_param_shift_rule_right, exp_param_shift_rule_right)
+        valid &= is_left_valid and is_right_valid
+    return valid
+
+
+def _has_grad_recipe(op):
+    """Checks whether grad_recipe is defined"""
+    if not hasattr(op, "grad_recipe"):
+        return False
+
+    if not any(map(lambda x: x, op.grad_recipe)):
+        return False
+
+    return True
+
+
+def _has_parameter_frequencies(op):
+    try:
+        if not hasattr(op, "parameter_frequencies"):
+            return False
+    except qml.operation.ParameterFrequenciesUndefinedError:
+        return False
+    return True
```

Comment:
`hasattr` should not error right?
```suggestion
    return hasattr(op, "parameter_frequencies")
```
Or we need to attempt to access it as well?
```suggestion
    if not hasattr(op, "parameter_frequencies"):
        return False
    try:
        op.parameter_frequencies
    except qml.operation.ParameterFrequenciesUndefinedError:
        return False
    return True
```

Reply:
```diff
diff --git a/frontend/catalyst/device/op_support.py b/frontend/catalyst/device/op_support.py
index 8d619eac..94029436 100644
--- a/frontend/catalyst/device/op_support.py
+++ b/frontend/catalyst/device/op_support.py
@@ -93,6 +93,7 @@ def _has_parameter_frequencies(op):
         if not hasattr(op, "parameter_frequencies"):
             return False
     except qml.operation.ParameterFrequenciesUndefinedError:
+        breakpoint()
         return False
     return True
 
```

```
pytest frontend/test/pytest/test_gradient.py --pdb

frontend/test/pytest/test_gradient.py ...................................................................................xx................................................................XXXXxxxxx.xx.....................
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/ubuntu/Code/catalyst/frontend/catalyst/device/op_support.py(97)_has_parameter_frequencies()
-> return False
(Pdb) 
```

Reply:
That's wild 👀